spark-cassandra-connector/src/main/scala/com/datastax/spark/connector/rdd/partitioner/CassandraRDDPartitioner.scala

```diff
@@ -109,14 +107,17 @@ class CassandraRDDPartitioner[V, T <: Token[V]](
             s"not supported in where. Use filter instead.")
     }.toSet
 
-    if (whereColumns.nonEmpty && whereColumns.size < pk.size) {
+    val primaryKeyComplete = whereColumns.nonEmpty && whereColumns.size == pk.size
+    val whereColumnsAllIndexed = whereColumns.forall(tableDef.isIndexed)
+
+    if (!primaryKeyComplete && !whereColumnsAllIndexed) {
       val missing = pk -- whereColumns
       throw new UnsupportedOperationException(
         s"Partition key predicate must include all partition key columns. Missing columns: ${missing.mkString(",")}"
```

Comment:
probably want to change this message to also include "Or all where predicates must be indexed"